enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+	// For route based setting of headers
+	SetHTTPResponseHeaders              []SetHTTPHeader
+	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
+}
+```
+```go
+// SetHTTPHeader specifies an HTTP header that should be set.
+type SetHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+
+	// Value specifies the header value.
+	Value string
+
+	// Action specifies the action to be performed.
+	Action string
+}
+
+type DeleteHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
+on either `spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` which will have the desired
+sequence.
+
+C) We will be creating an immutable/harcoded deny list of HTTP Response headers in the [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141) 
+which won't be allowed to be set/delete. Eg - HSTS.
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Would this enhacement cover setting request headers ?
+
+  Answer: As per [epic](https://issues.redhat.com/browse/RFE-464) the requirement for the customers seems to be for HTTP response headers only.
+  We need confirmation from the PM regarding this.
+
+- Would we be able to set load balancing alogorithm based on headers ?
+
+  Answer: No, this out of scope for this enhancement so it will be considered as a non-goal.
+
+- Would we be able to delete or append headers  ?
+
+  Answer: Yes, this enhancement will cover deletion and but not appending of HTTP Response Headers as per the [epic](https://issues.redhat.com/browse/RFE-464). 
+
+- Would some headers be blocked for setting and deleting ?
+
+  Answer: Yes, this enhancement won't allow the user to set or delete HSTS headers. We already a way to set this which is via route 
+  annotation. Providing one more way to add this can cause issues.
+
+- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)?
+  Answer: No, this is out of scope for this enhancement.
+
+- Would we be able to [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) ?
+  Answer:  No, this is out of scope for this enhancement.
```

Comment:
Could these be future extensions or do you not see this ever being supported?

Reply:
The epic does not demand it so we are not implementing this in this release. However the API can be extended by adding  a replace action.

Reply:
How much extra work would it be to implement this now, in the initial implementation, vs later?

Reply:
not sure.